sql: allow CREATE UNLOGGED TABLE syntax

otan opened the pull request:
In SQL, a CREATE TABLE can take in TEMP | UNLOGGED. Accordingly, I've
updated the sql.y to account for this, changing the opt_temp from a bool
to a `tree.Persistence` ENUM that can be PERMANENT, TEMPORARY or
UNLOGGED. This introduced some downstream changes to various package to
recognize these arguments. Note sequences and views cannot be unlogged.
This makes up the majority of the changes.

Otherwise, recognize UNLOGGED TABLE in teh parser. Add a warning that the
use of UNLOGGED TABLE does not actually do anything.
Note this is always safe as UNLOGGED TABLE is a performance improvement
in PostgreSQL by sacrificing crash-safety. In our case, we'll still always
have safety features and thus still be correct, albeit slower.

Release note (sql change): Allow `CREATE UNLOGGED TABLE` syntax to be
parsed and recorded internally within CockroachDB. However, this syntax
does not modify behavior of the table.

Resolves #43352.
Refs #51818

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52596)
<!-- Reviewable:end -->

rohany commented:
If it does nothing, why are you storing it on the table descriptor?

otan commented:
I don't know where to stop! 

I may have to implement SET UNLOGGED and SET LOGGED and it needs that to be set to behave correctly, hehe.

rohany commented:
I don't think we should be writing this information to disk if it is to be used just to play with these commands. Why can't we just parse and noop the related commands?

otan commented:
i honestly could be swayed either way.  

i don't want a no-op as i do want an explicit warning/notice, which you cannot do from the sql.y layer.

rohany commented:
I agree that you'll have to implement some plan nodes, but I'd prefer for those to be noops than having to do something to the table

otan commented:
alright well, change made :P

mgartner commented on pkg/sql/show_create.go:
```diff
@@ -73,6 +73,9 @@ func ShowCreateTable(
 	f.WriteString("CREATE ")
+	if desc.Unlogged {
+		f.WriteString("UNLOGGED ")
+	}
```
I agree with @rohany. If the `UNLOGGED` keyword does nothing, I think it's misleading to label the table as `UNLOGGED` in `SHOW CREATE`. If you remove this, I think you can avoid storing anything new in a descriptor, no?

otan replied:
thought i removed this on cleanup!

otan committed:
sql: allow CREATE UNLOGGED TABLE syntax

In SQL, a CREATE TABLE can take in TEMP | UNLOGGED. Accordingly, I've
updated the sql.y to account for this, changing the opt_temp from a bool
to a `tree.Persistence` ENUM that can be PERMANENT, TEMPORARY or
UNLOGGED. This introduced some downstream changes to various package to
recognize these arguments. Note sequences and views cannot be unlogged.
This makes up the majority of the changes.

Otherwise, recognize UNLOGGED TABLE in teh parser. Add a warning that the
use of UNLOGGED TABLE does not actually do anything.
Note this is always safe as UNLOGGED TABLE is a performance improvement
in PostgreSQL by sacrificing crash-safety. In our case, we'll still always
have safety features and thus still be correct, albeit slower.

Release note (sql change): Allow `CREATE UNLOGGED TABLE` syntax to be
parsed and recorded internally within CockroachDB. However, this syntax
does not modify behavior of the table.

otan commented:
thankssss

bors r=rohany

otan commented:
bors bors the magical fruit
the more you eat the more you toot
(didn't seem to want to process my PR)

bors r=rohany

dt commented:
bors? flaky? unheard of.
